<font size="4">
<center><b>Анализ базы данных книг</b></center>
</font>

**Описание проекта:** Компания купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.  

**Цель исследования:** Проанализировать базу данных с помощью SQL-запросов и сформулировать ценностное предложение для нового продукта.

**Описание данных:** 

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Подключение библиотек и обзор данных

In [52]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

#### Таблица books

In [53]:
# смотрим первые 10 записей в таблице
request = """
SELECT *
FROM books
LIMIT 5;
"""
pd.io.sql.read_sql(request, con = engine)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [54]:
# узнаем число записей в таблице
request = """
SELECT COUNT(*)
FROM books;
"""
pd.io.sql.read_sql(request, con = engine)

count
0   1000

#### Таблица authors

In [55]:
# смотрим первые 10 записей в таблице
request = """
SELECT *
FROM authors
LIMIT 5;
"""
pd.io.sql.read_sql(request, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [56]:
# узнаем число записей в таблице
request = """
SELECT COUNT(*)
FROM authors;
"""
pd.io.sql.read_sql(request, con = engine)

count
0    636

#### Таблица publishers

In [57]:
# смотрим первые 10 записей в таблице
request = """
SELECT *
FROM publishers
LIMIT 5;
"""
pd.io.sql.read_sql(request, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [58]:
# узнаем число записей в таблице
request = """
SELECT COUNT(*)
FROM publishers;
"""
pd.io.sql.read_sql(request, con = engine)

count
0    340

#### Таблица ratings

In [59]:
# смотрим первые 10 записей в таблице
request = """
SELECT *
FROM ratings
LIMIT 5;
"""
pd.io.sql.read_sql(request, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [60]:
# узнаем число записей в таблице
request = """
SELECT COUNT(*)
FROM ratings;
"""
pd.io.sql.read_sql(request, con = engine)

count
0   6456

#### Таблица reviews

In [61]:
# смотрим первые 10 записей в таблице
request = """
SELECT *
FROM reviews
LIMIT 5;
"""
pd.io.sql.read_sql(request, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [62]:
# узнаем число записей в таблице
request = """
SELECT COUNT(*)
FROM reviews
"""
pd.io.sql.read_sql(request, con = engine)

count
0   2793

### Сколько книг вышло после 1 января 2000 года?

In [63]:
request = """
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
"""
pd.io.sql.read_sql(request, con = engine)

count
0    819

После 01.01.2000 выпущено 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [64]:
request = """
SELECT b.title,
COUNT(DISTINCT rev.review_id),
AVG(ret.rating)
FROM books AS b
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS ret ON b.book_id = ret.book_id
GROUP BY b.book_id
ORDER BY avg DESC,
count DESC;
"""
pd.io.sql.read_sql(request, con = engine)

title  count   avg
0                        A Dirty Job (Grim Reaper  #1)      4  5.00
1              School's Out—Forever (Maximum Ride  #2)      3  5.00
2         Moneyball: The Art of Winning an Unfair Game      3  5.00
3        Arrows of the Queen (Heralds of Valdemar  #1)      2  5.00
4    Wherever You Go  There You Are: Mindfulness Me...      2  5.00
..                                                 ...    ...   ...
995  The World Is Flat: A Brief History of the Twen...      3  2.25
996                                      Drowning Ruth      3  2.00
997                  His Excellency: George Washington      2  2.00
998                                              Junky      2  2.00
999                               Harvesting the Heart      2  1.50

[1000 rows x 3 columns]

Книга с самым большим количеством обзоров и наивысшим средним рейтингом- A Dirty Job (Grim Reaper). Самый низкий средний рейтинг у книги Harvesting the Heart - 1,5.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [65]:
request = """
SELECT p.publisher,
COUNT(b.book_id)
FROM publishers AS p
LEFT JOIN books AS b ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id
ORDER BY count DESC
LIMIT 1;
"""
pd.io.sql.read_sql(request, con = engine)

publisher  count
0  Penguin Books     42

Издательство Penguin Books выпустило 42 книги, и это наибольший показатель среди всех издательств.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [66]:
request = """
SELECT author, AVG(rating)
FROM authors AS a
JOIN books AS b 
ON a.author_id = b.author_id
JOIN ratings AS r 
ON b.book_id = r.book_id

WHERE b.book_id IN (
SELECT book_id
FROM ratings 
GROUP BY book_id
HAVING COUNT(rating_id) >= 50
ORDER BY COUNT(rating_id) DESC)
GROUP BY author
ORDER BY avg DESC
LIMIT 1

"""
pd.io.sql.read_sql(request, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

Самая высокая средняя оценка книги - 4,28. Автором, написавшим эту книгу является J.K.Rowling/Mary GrandPré

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [67]:
request = """
SELECT ROUND(AVG(count)) 
FROM
(SELECT COUNT(text)
FROM reviews
FULL OUTER JOIN (SELECT username,
COUNT(rating_id) as cnt
FROM ratings
GROUP BY username) AS cnt_rating
ON cnt_rating.username = reviews.username
WHERE cnt > 50
GROUP BY reviews.username) as count

"""
pd.io.sql.read_sql(request, con = engine)

round
0   24.0

Пользователи, которые поставили более 50 оценок, в среднем выпускаю 24 обзора.

### Общий вывод

<font size="3">
<br>- Сервис для чтения хранит 1000 различных книг от 636 авторов
<br>- из них 819 книг выпустились в 21ом веке
<br>- средняя оценка книг варьируется от 1,5 до 5
<br>- издательство Penguin Books выпустило наибольшее количество книг из данной коллекции
<br>- автором с самой высокой оценкой книги является J.K.Rowling/Mary GrandPré
<br>- самые активные пользователи сервиса выпускают в среднем 24 обзора.
    </font>